pkgs/buildNpmPackage/default.nix

```diff
@@ -0,0 +1,58 @@
+{ stdenvNoCC, writeShellScriptBin, writeText, stdenv, fetchurl, makeWrapper, nodejs-10_x }:
+let
+  inherit (builtins) fromJSON toJSON readFile split head elemAt foldl';
+
+  deps-to-fetches = base: deps: builtins.foldl' (a: b: a // (dep-to-fetch b)) base (builtins.attrValues deps);
+  dep-own-fetch = { resolved, integrity, ... }: let
+    ssri = split "-" integrity;
+    hashtype = head ssri;
+    hash = elemAt ssri 2;
+  in {
+    "${resolved}" = fetchurl {
+      url = resolved;
+      "${hashtype}" = hash;
+    };
+  };
+
+  dep-to-fetch = args @ { resolved ? null, dependencies ? {}, ... }:
+    deps-to-fetches (if isNull resolved then {} else dep-own-fetch args) dependencies;
+  npm-cache-input = lock: writeText "npm-cache-input.json" (toJSON (dep-to-fetch lock));
+
+  patchShebangs = writeShellScriptBin "patchShebangs.sh" ''
+    set -e
+    source ${stdenvNoCC}/setup
+    patchShebangs "$@"
+  '';
+  shellWrap = writeShellScriptBin "npm-shell-wrap.sh" ''
+    set -e
+    if [ ! -e .shebangs_patched ]; then
+      ${patchShebangs}/bin/patchShebangs.sh .
+      touch .shebangs_patched
+    fi
+    exec bash "$@"
+  '';
+in
+args @ { lockfile, src, buildInputs ? [], ... }:
+let lock = fromJSON (readFile lockfile); in
+stdenv.mkDerivation ({
+  inherit (lock) version;
+  name = "${lock.name}-${lock.version}";
+
+  XDG_CONFIG_DIRS = ".";
+  NO_UPDATE_NOTIFIER = true;
+  buildPhase = ''
+    echo making cache
+    node ${./mkcache.js} ${npm-cache-input lock}
+    echo installing
+    npm ci --cache=./npm-cache --offline --script-shell=${shellWrap}/bin/npm-shell-wrap.sh
```

Comment:
I was about to tell you to use full command names instead of abbreviations in scripts, but it turns out `npm ci` *is* the full command name. Cue traditional npm-inspired facepalm.